tsl/src/remote/dist_copy.c

```diff
@@ -1366,12 +1370,35 @@ remote_copy_process_and_send_data(RemoteCopyContext *context)
 		else
 			found = true;
 
-		/* get the filtered list of "available" DNs for this chunk but only if it's replicated */
-		if (found && ht->fd.replication_factor > 1)
+		/*
+		 * Get the filtered list of "available" DNs for this chunk but only if it's replicated. We
+		 * only fetch the filtered list once. Assuming that inserts will typically go to the same
+		 * chunk we should be able to reuse this filtered list a few more times
+		 *
+		 * The worse case scenario is one in which INSERT1 goes into CHUNK1, INSERT2 goes into
+		 * CHUNK2, INSERT3 goes into CHUNK1,... in which case we will end up refreshing the list
+		 * everytime
+		 *
+		 * We will also enter the below loop if we KNOW that any of the DNs has been marked
+		 * unavailable before we started this transaction. If not, then we know that every chunk's
+		 * datanode list is fine and no stale chunk metadata updates are needed.
+		 */
+		if (context->dns_unavailable && found && ht->fd.replication_factor > 1)
```

Comment:
Shouldn't we make a call to reset `context->dns_unavailable` for the remaining rows in the batch after removing the data nodes for stale chunk?

Reply:
No. The `dns_unavailable` is a global state for the entire transaction. We just want to avoid doing these expensive checks if all DNs are available.